### ISSUE TITLE
no-magic-numbers allow 1 number per declaration

### ISSUE THREAD

**Issue by keithamus** [eslint/eslint#5396]
> This comment was left in https://github.com/eslint/eslint/issues/4226 but I think as its closed its probably worth filing a new issue, especially as it is somewhat tangential to the original issue.

I would request this rule for instances where a simple calculation is needed. Take for example the following code:

```js
const myArray = [1,2,3];
const lastIndex = myArray.length - 1;
```
This const uses 1 number as part of the calculation, and should be allowed to be passed through - because the number on its own has no semantic value, and trying to give it a useful name becomes a struggle, e.g:

```js
const myArray = [1,2,3];
const lastIndexLengthDelta = 1;
const lastIndex = myArray.length - lastIndexLengthDelta;
```
I'd argue the first example, while not attaching a semantic value to the number, is more readable than the 2nd.

### proposal 

I understand that I can just whitelist the number 1, or two - but I think it would be even better have the option of a threshold of numbers used in a calculation. E.g.
```js
/* eslint no-magic-numbers: 2 { "numbersAllowedInCalculations": 0 } */
const lastIndex = myArray.length - 1; // errors

/* eslint no-magic-numbers: 2 { "numbersAllowedInCalculations": 1 } */
const lastIndex = myArray.length - 1; // does not error
```

**Comment by ilyavolodin** [comment]
So are you saying that this pattern shouldn't be flagged with proposed option:

```js
const lastIndex = myArray.length - 132;
```

**Comment by keithamus** [comment]
@ilyavolodin I am saying that yes. I understand that you're trying to point out that `132` is a magic number, but consider the counter examples;

```js
const lastIndex = myArray.length - 1;
const halfElementWidth = element.outerWidth / 2;
const defaultTimeout = 30 * oneSecondInMs;
```

Are any of these benefited by having extra constants?

```js
const lastIndexLengthDelta = 1;
const lastIndex = myArray.length - lastIndexLengthDelta;

const halfDivider = 2;
const halfElementWidth = element.outerWidth / halfDivider;

const theNumberThirty = 30;
const defaultTimeout = theNumberThirty * oneSecondInMs;
```

These numbers bear no meaning outside of the relation they have to the calculation - they're not constant values which could be tweaked, they're not used outside of the single calculation, and changing their value will very likely change any semantic value they have (e.g. changing `halfDivider` to `3` suddenly makes it `thirdDivider` and `halfElementWidth` becomes `thirdElementWidth`).

**Comment by ilyavolodin** [comment]
@keithamus Not going to argue about first two examples. That's why when this rule was implemented originally, it excluded 1 and 2 by default. It was change later so that people could configure those numbers they wanted to be excluded.
Third example, though could (and in my mind, should) be rewritten as:

```js
const defaultTimeoutInSeconds = 30;
const defaultTimeout = defaultTimeoutInSeconds * oneSecondInMs;
```

I'll defer to other members of @eslint/eslint-team but you haven't convinced me yet.

**Comment by BYK** [comment]
I'm :-1:. A number without an explanation is a magic number to me. And explaining means assigning it to a constant.

**Comment by michaelficarra** [comment]
Definitely :-1: on this. The examples that were supposed to support this change are actually great examples of why this rule was created.

**Comment by nzakas** [comment]
@keithamus sorry, it doesn't seem there's enough support to make a change here.

**Comment by keithamus** [comment]
Fair enough. I'll take the downvotes on board and try to find suitable ways to use this in future.
